e2e/exec.go

```diff
@@ -97,29 +98,80 @@ func sshKeyName(vmPrivateIP string) string {
 }
 
 func sshString(vmPrivateIP string) string {
-	return fmt.Sprintf(`ssh -i %s -o PasswordAuthentication=no -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o ConnectTimeout=5 azureuser@%s`, sshKeyName(vmPrivateIP), vmPrivateIP)
+	return fmt.Sprintf(`ssh -i %[1]s -o PasswordAuthentication=no -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o ConnectTimeout=5 azureuser@%[2]s`, sshKeyName(vmPrivateIP), vmPrivateIP)
 }
 
-func execOnVM(ctx context.Context, kube *Kubeclient, vmPrivateIP, jumpboxPodName, sshPrivateKey, command string) (*podExecResult, error) {
-	sshCommand := fmt.Sprintf(`echo '%s' > %[2]s && chmod 0600 %[2]s && %s`, sshPrivateKey, sshKeyName(vmPrivateIP), sshString(vmPrivateIP))
-	sshCommand = sshCommand + " sudo"
-	commandToExecute := fmt.Sprintf("%s %s", sshCommand, command)
+func quoteForBash(command string) string {
+	return fmt.Sprintf("'%s'", strings.ReplaceAll(command, "'", "'\"'\"'"))
+}
+
+func execBashCommandOnVM(ctx context.Context, s *Scenario, vmPrivateIP, jumpboxPodName, sshPrivateKey, command string) (*podExecResult, error) {
+	image := &config.Image{
+		OS: config.OSUbuntu,
+	}
+	return execScriptOnVm(ctx, s, vmPrivateIP, jumpboxPodName, sshPrivateKey, []string{command}, image)
+}
+
+func execScriptOnVm(ctx context.Context, s *Scenario, vmPrivateIP, jumpboxPodName, sshPrivateKey string, script []string, os *config.Image) (*podExecResult, error) {
+	/*
+		This works in a way that doesn't rely on the node having joined the cluster:
+		* We create a linux pod on a different node.
+		* on that pod, we create a script file containing the script passed into this method.
+		* Then we scp the script to the node under test.
+		* Then we execute the script using an interpreter (powershell or bash) based on the OS of the node.
+	*/
+	identifier := uuid.New().String()
+
+	var scriptFileName, remoteScriptFileName string
+	if os.OS == config.OSWindows {
+		scriptFileName = fmt.Sprintf("script_file_%s.ps1", identifier)
+		remoteScriptFileName = fmt.Sprintf("c:/%s", scriptFileName)
+	} else {
+		scriptFileName = fmt.Sprintf("script_file_%s.sh", identifier)
+		remoteScriptFileName = scriptFileName
+	}
+
+	var interpreter string
+	switch os.OS {
+	case config.OSWindows:
+		interpreter = "powershell"
+		break
+	default:
+		interpreter = "bash"
+		break
+	}
+
+	scriptWithLineBreaks := strings.Join(script, "\n")
+	steps := []string{
+		fmt.Sprintf("echo '%[1]s' > %[2]s", sshPrivateKey, sshKeyName(vmPrivateIP)),
+		"set -x",
+		fmt.Sprintf("echo %[1]s > %[2]s", quoteForBash(scriptWithLineBreaks), scriptFileName),
+		fmt.Sprintf("chmod 0600 %s", sshKeyName(vmPrivateIP)),
+		fmt.Sprintf("chmod 0755 %s", scriptFileName),
```

Comment:
do you need to chmod after scp or before?

Reply:
I think scp maintains permissions so that's why I chmod before - otherwise it's a seperate command to change the permissions before executing it. Although we're specifying the interpreter so the chmod is probably unnecessary anyway.